### ISSUE TITLE
Non-local cli tools

### ISSUE THREAD

**Issue by thusoy** [havardgulldahl/jottalib#16]
I've started a fork working on a couple of cli tools for working with the jottacloud folder, like `jotta-ls`, `jotta-rm`, `jotta-upload` and `jotta-download` (or maybe just `jotta-cp`), etc. My intention is to be able to manipulate the data without needing a full copy locally, or needing to mount a filesystem or anything like that. I think I also got big file handling working, but I couldn't get it to work while still utilizing `requests_cache`. Should I propose a pull request to get these tools into jottalib when they're closer to completion, or do you think they should live in a separate project, like `jottacloud-cli` or similar?

Great work on getting this project going by the way! :+1: Sad that Jotta haven't seen any value in it yet.

**Comment by jkaberg** [comment]
@thusoy Would you mind rebasing and sending a PR to @havardgulldahl branch?

**Comment by havardgulldahl** [comment]
@thusoy Sorry about the lag. I'm intrigued by your tools, what kind of use
cases are you solving?

Re the big file handling, I'll take a look at your solution.

And I'd like to keep the jottalib repository as clean as possible.. maybe
rename jottacloudclient to -tools and put everything in there? What do you
guys think?

On Mon, May 25, 2015 at 6:11 PM, Joel Kåberg <notifications@github.com>
wrote:

> @thusoy <https://github.com/thusoy> Would you mind rebasing and sending a
> PR to @havardgulldahl <https://github.com/havardgulldahl> branch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/havardgulldahl/jottalib/issues/16#issuecomment-105258140>
> .
>



-- 
Håvard Gulldahl <havard@gulldahl.no>
Telefon: 9971 0615
http://lurtgjort.no/

**Comment by havardgulldahl** [comment]
I've had another think about this, and actually what we'll do is put all cli tools, jottacloudclient included, into the jottalib repository. 

And then we'll have different packages on pypi, all created from one repo. Everything else seems really messy to me, right now. So go ahead and create that PR once you're done, @thusoy :)

**Comment by thusoy** [comment]
@jkaberg Sure, glad to see the project have picked up some steam again.

@havardgulldahl As FUSE and duplicity are not great friends of windows, it's a way to get cross-platform shell access to your files. I'm using duplicity for most of the backups on other machines, but it's nice to be able to do a sanity-check of the backups without running duplicity in full, just checking that recent files are present, and for pushing some non-backup files up in the cloud for sharing or temporary storage.

I'll clean it up and submit a PR come summer, got a thesis that needs to be completed first.

**Comment by havardgulldahl** [comment]
I see. Thanks for the explanation. 

Are you using  duplicity with the jottacloud backend, or with jottafuse as a "local" backend?

**Comment by thusoy** [comment]
@havardgulldahl Duplicity with Jottacloud backend, with some modifications to handle large files. Sadly, duplicity itself is less than great with large files...

**Comment by havardgulldahl** [comment]
Any news on your fork, @thusoy? I'm very intrigued by your cli tools and the reorganization of scripts that you have done, and I'd like to pull it in and build on top of that for the next release. 

I realize that theses have their own timeline, and if you have some ballpark schedule for this, I'd be happy. :)

**Comment by thusoy** [comment]
I didn't find as much time for this during summer as hoped,  I'm out travelling now and will be for the next few weeks,  ballpark is late august. It's not much work really,  so should be done quickly as soon as I'm back. 

Den 30. juli 2015 21.53.11 CEST, skrev "Håvard Gulldahl" <notifications@github.com>:

**Comment by havardgulldahl** [comment]
@thusoy How's it going? Let's get your stuff into the tree!


And, I'd be really happy if you could explain your `duplicity` setup over here: https://github.com/havardgulldahl/jottalib/wiki/Normal-use-cases#i-want-to-integrate-with-the-duplicity-backup-program

**Comment by thusoy** [comment]
I rebased and opened a PR, not ready for merge yet but feedback welcome. I can split out the not directly related commits to a different PR if desired (editorconfig and gitignore), the rest I'll squash before merge.

**Comment by thusoy** [comment]
As for the duplicity integration, I think I started from the duplicity backend included in this repo and ended up here: https://gist.github.com/thusoy/def300d58b034da0ba91. I unpacked duplicity 0.7 and put this file in `duplicity/backends/jotta_backend.py`, then ran `python setup.py install`. Has been working pretty good so far. This has been working pretty steady now for several months with `jottalib` installed from my cli branch. Only issue have been a couple of memory errors on a host with a pretty big backup set, but I think those are from duplicity core, not the jottacloud integration.

**Comment by havardgulldahl** [comment]
Thank you for that gist. I added your duplicity enhancements to `master`.

I'll comment on your PR over there: #50

**Comment by thusoy** [comment]
@havardgulldahl Did some work on this today, split the PR up into clearly separated commits so that it's easy to follow what's done. The only thing I think is really missing is a progress bar for uploads. Downloads has one. I see that in some of the tests you've used `clint` for this, but since it hadn't been added as a requirement yet I wasn't sure if you wanted to keep it out or not, so I just wrote what was needed for the download progress myself.

As you can see from updated PR, many of the commits are not directly related to the CLI tools. Do you want me to split these to separate PRs or will you just merge the entire bunch?

You can also see here that I refactored auth handling to use the netrc file that you mentioned, and added a single function for this that's now used by all the tools that previously handled it themselves.

**Comment by thusoy** [comment]
Added progress bar for uploads as well. Give it a spin, everything should be working now.

**Comment by havardgulldahl** [comment]
@thusoy Ok, so most of this looks straight forward and obviously an enhancement. This is great! 

The only thing is progressbar.py. 

I'd really like to continue using clint. I have every confidence in your progressbar implementation, but since we're using clint for colored output anyway, and less code to maintain == less bugs, I suggest we use the clint progressbar all over the place. 

What are the downsides to this, as you see it?

**Comment by havardgulldahl** [comment]
Oh, and I think we should add `jotta-scan` and `jotta-monitor` in there as well. To get rid of the ridiculous names they're carrying now.

**Comment by thusoy** [comment]
@havardgulldahl No objections at all to using clint, I just saw that it was only used in tests and not in any of the shipped code, and thus I didn't want to add an extra dependency just for the progressbar. But I have no emotional attachment to my own code, I'll delete that and add `clint` as a dependency and use that instead, gimme a sec.

**Comment by havardgulldahl** [comment]
sweet.

**Comment by thusoy** [comment]
@havardgulldahl Done. However, I do think it's a step down in terms of usability - clint doesn't display the current download/upload speed as the previous code did, only <number of bytes downloaded>/<total size>, which doesn't make much sense to most users. I'll see if I can subclass the clint progressbar to show speed instead of byte count, but I can provide that in a separate PR if you're satisfied with this one. :)

**Comment by thusoy** [comment]
Turns out it was hard to subclass the clint progress bar. I'll submit a patch to clint and see if I can get a progress subclass for file transfer into their tree, and if that happens we can use that here without getting any extra maintenance.

**Comment by thusoy** [comment]
Ah, now I see what you mean with the progressbar, that file was commited by accident. I'll remove it from the PR as soon as I get home.

**Comment by thusoy** [comment]
Unused progress.py removed.
